docs/src/sir.jl

```diff
@@ -0,0 +1,134 @@
+# # Non-Markovian SIR Model
+
+using Random
+using Plots
+using Distributions
+using Fleck
+
+# In this tutorial we demonstrate construction of a SIR (susceptible-infectious-removed) model
+# with non-exponential recovery times. Infection events occur at the points of a Poisson process,
+# which is equivalent to using a single exponential clock whose rate corresponds to the overall
+# rate of infection in the population. The infection rate is ``\beta c S I / N`` where ``N`` is
+# the total population size, making this a *frequency-dependent* force of infection term rather
+# than pure mass action (although as ``N`` does not change in this example, the difference is moot).
+# Clocks for recovery events follow an arbitrary distribution.
+
+# ## Model structure
+
+# We use a struct that stores the type of the recovery distribution as a type parameter.
+# Additionally, because recovery clocks need unique keys, we define the method `get_key!`
+# which retrives an integer key and increments the stored key counter.
+
+# The `initialize!` method uses the initial state to enable the infection clock and
+# recovery clocks for each initial infectious person. Note that the keys are a tuple
+# where the first element is a `Symbol` giving the event type, and the second element
+# is an integer.
+
+mutable struct SIRNonMarkov{T<:Distribution}
+    state::Vector{Int}
```

Comment:
Have you thought about the pros and cons of writing the state as S::Int, I::Int, and R::Int? versus the vector? I went through this, and that was my one thought on clarity. Or, if not, say the state is a vector of length 3 representing S,I,R.

Reply:
Personally I think the best solution for a "real simulation" would be a statically sized array from https://github.com/JuliaArrays/StaticArrays.jl but I think these examples should depend on as few external packages as possible. I would rather have the vector, and just say that it represents S,I,R.

Reply:
resolved with https://github.com/adolgert/Fleck.jl/pull/55/commits/6fcf9492001e906bdb6f0698a1afb0bab44ce91c